market_analysis/views/default.py

```diff
@@ -0,0 +1,450 @@
+                             headers=headers)
+        else:
+            return {'error': "Username or Password Not Recognized"}
+    return {'error': ''}
+
+
+@view_config(route_name='logout')
+def logout(request):
+    """Log out a user."""
+    headers = forget(request)
+    return HTTPFound(request.route_url('login'), headers=headers)
+
+
+@view_config(route_name='api_error', renderer='templates/api_error.jinja2')
+def api_error(request):
+    """Display a page with an error msg when can't connect to api."""
+    return {}
+
+
+def format_dates(date_list):
+    """Truncate date format."""
+    ret_list = []
+    for date in date_list:
+        date = date[5:10]
+        ret_list.append(date)
+    return ret_list
+
+
+def prepare_daily_changes(daily_totals):
+    """Return a list of daily percentage changes."""
+    daily_change = []
+    for tot in daily_totals:
+        if daily_totals[0] > 0:
+            daily_change\
+                .append(round(((tot * 100 / daily_totals[0]) - 100), 5))
+        else:
+            daily_change.append(0)
+    return daily_change
+
+
+def query_shares(request, user_id, symbol):
+    """Return shares for current user and stock symbol."""
+    current_stock_id = request.dbsession.query(Stocks)\
+        .filter(Stocks.symbol == symbol).first().id
+    shares = request.dbsession.query(Association).\
+        filter(and_(Association.stock_id == current_stock_id,
+               Association.user_id == user_id)).first().shares
+    return int(shares)
+
+
+def build_stock_entry(
+        y_values, price, shares, value, max_num=None, min_num=None
+        ):
+    """Build a stock entry to be returned to build graph."""
+    return {'y_values': y_values, 'price': price, 'shares': shares,
+            'value': value, 'max': max_num, 'min': min_num}
+
+
+def build_graph(request, elements, percentage=False):
+    """Build and return the graph data from an API request for rendering."""
+    url = 'http://dev.markitondemand.com/MODApis/Api/v2/InteractiveChart/json'
+    req_obj = {     # parameters for API request
+        "parameters":
+        {
+            'Normalized': 'false',
+            'NumberOfDays': 14,
+            'DataPeriod': 'Day',
+            'Elements': elements
+        }
+    }
+    resp = requests.get(url, params=urlencode(req_obj))
+
+    if resp.status_code == 200:
+        total_shares = 0
+        total_value = 0
+        entries = {key: value for key, value in resp.json().items()}    # data from API
+        stocks = {}     # data entries for each stock
+        export = {}     # container of re-packaged data to be rendered
+
+    # build export dict for template
+        export['dates'] = format_dates(entries['Dates'])
+        export['x_values'] = entries['Positions']
+        daily_totals = [0 for j in range(len(export['x_values']))]
+
+        current_user_id = request.dbsession.query(Users).filter(
+            Users.username == request.authenticated_userid
+        ).first().id
```

Comment:
Everywhere you use this `build_graph` function, you already have acquired the authenticated user's db id.  You could just pass that in instead of doing another query for the same data inline inside this method.